File: Examples/ObjC/Sources/ContentCards-Custom-UI/CardsInfoViewController.m
#import "CardsInfoViewController.h"

@import UIKit;
@import BrazeKit;

#pragma mark - Field / Section

@interface Field : NSObject

@property(copy, nonatomic) NSString *name;
@property(strong, nonatomic) id value;
@property(assign, nonatomic) NSInteger indentation;

+ (instancetype)fieldWithName:(NSString *)name value:(id)value;
+ (instancetype)fieldWithName:(NSString *)name
                        value:(id)value
                  indentation:(NSInteger)indentation;

@end

@interface Section : NSObject

@property(copy, nonatomic) NSString *name;
@property(strong, nonatomic) NSArray<Field *> *fields;

@end

#pragma mark - CardsInfoViewController

@interface CardsInfoViewController ()

@property(strong, nonatomic) Braze *braze;
@property(strong, nonatomic) NSArray<BRZContentCardRaw *> *cards;
@property(strong, nonatomic) NSArray<Section *> *sections;
// Must be retained to keep subscription active; cancelling re-applies local analytics.
@property(strong, nonatomic) BRZCancellable *subscription;
// Tracks cards already impressed in this presentation to avoid duplicate impression events.
@property(strong, nonatomic) NSMutableSet<NSString *> *impressedCardIDs;

+ (Section *)cardSectionFromCard:(BRZContentCardRaw *)card
                         atIndex:(NSInteger)index;

@end

@implementation CardsInfoViewController

- (instancetype)initWithBraze:(Braze *)braze {
  self = [super initWithStyle:UITableViewStyleGrouped];
  if (self) {
    self.braze = braze;
    self.cards = @[];
    self.sections = @[];
    self.impressedCardIDs = [NSMutableSet set];

    self.title = @"Content Cards Info";

    UIBarButtonItem *doneButton = [[UIBarButtonItem alloc]
        initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                             target:self
                             action:@selector(dismissModal)];
    [self.navigationItem setRightBarButtonItem:doneButton];

    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 44.0;
  }
  return self;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  // Subscribe to content cards updates. Delivers an initial snapshot of cached cards
  // immediately, then subsequent updates as cards change. Cancelling re-applies local
  // analytics (viewed/clicked/removed) and notifies any remaining subscribers.
  // Note: impressions, clicks, and dismissals do not trigger the subscription callback
  __weak typeof(self) weakSelf = self;
  self.subscription = [self.braze.contentCards
      subscribeToUpdates:^(NSArray<BRZContentCardRaw *> *cards) {
        typeof(self) strongSelf = weakSelf;
        if (!strongSelf) return;
        strongSelf.cards = cards;
        NSMutableArray *sections = [NSMutableArray array];
        for (NSInteger i = 0; i < cards.count; ++i) {
          [sections addObject:[CardsInfoViewController cardSectionFromCard:cards[i]
                                                                   atIndex:i]];
        }
        strongSelf.sections = [sections copy];
        [strongSelf.tableView reloadData];
      }];
}

- (void)dismissModal {
  [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return self.sections.count;
}

- (NSInteger)tableView:(UITableView *)tableView
    numberOfRowsInSection:(NSInteger)section {
  return self.sections[section].fields.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell =
      [tableView dequeueReusableCellWithIdentifier:@"cellIdentifier"];
  if (cell == nil) {
    cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1
                                  reuseIdentifier:@"cellIdentifier"];
  }

  Field *field = self.sections[indexPath.section].fields[indexPath.row];
  cell.textLabel.text = field.name;
  cell.detailTextLabel.text = [NSString stringWithFormat:@"%@", field.value];
  cell.indentationLevel = field.indentation;

  cell.textLabel.numberOfLines = 0;
  if (@available(iOS 13.0, tvOS 13.0, *)) {
    cell.textLabel.font =
        [UIFont monospacedSystemFontOfSize:13 weight:UIFontWeightRegular];
    cell.detailTextLabel.font =
        [UIFont monospacedSystemFontOfSize:13 weight:UIFontWeightRegular];
  }

  return cell;
}

#pragma mark - UITableViewDelegate

- (NSString *)tableView:(UITableView *)tableView
    titleForHeaderInSection:(NSInteger)section {
  return self.sections[section].name;
}

- (UIView *)tableView:(UITableView *)tableView
    viewForFooterInSection:(NSInteger)section {
  UIButton *dismissButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [dismissButton setTitle:@"Log Dismissal" forState:UIControlStateNormal];
  dismissButton.tag = section;
  [dismissButton addTarget:self
                    action:@selector(logDismissal:)
          forControlEvents:UIControlEventTouchUpInside];

  UIButton *clickButton = [UIButton buttonWithType:UIButtonTypeSystem];
  [clickButton setTitle:@"Log Click" forState:UIControlStateNormal];
  clickButton.tag = section;
  [clickButton addTarget:self
                  action:@selector(logClick:)
        forControlEvents:UIControlEventTouchUpInside];

  UIStackView *stack = [[UIStackView alloc]
      initWithArrangedSubviews:@[ clickButton, dismissButton ]];
  stack.axis = UILayoutConstraintAxisVertical;
  stack.spacing = 8;

  UIView *container = [[UIView alloc] init];
  [container addSubview:stack];
  stack.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [stack.centerXAnchor constraintEqualToAnchor:container.centerXAnchor],
    [stack.topAnchor constraintEqualToAnchor:container.topAnchor constant:8],
    [stack.bottomAnchor constraintEqualToAnchor:container.bottomAnchor
                                        constant:-8],
  ]];
  return container;
}

// Log an impression when the first row of each card section becomes visible.
// Deduplication prevents a second impression if the card scrolls off-screen and back.
- (void)tableView:(UITableView *)tableView
    willDisplayCell:(UITableViewCell *)cell
    forRowAtIndexPath:(NSIndexPath *)indexPath {
  if (indexPath.row != 0) return;
  BRZContentCardRaw *card = self.cards[indexPath.section];
  if ([self.impressedCardIDs containsObject:card.identifier]) return;
  [self.impressedCardIDs addObject:card.identifier];
  [card.context logImpression];
}

- (void)logClick:(UIButton *)sender {
  BRZContentCardRaw *card = self.cards[sender.tag];
  [card.context logClick];
  if (card.url != nil) {
    [card.context processClickActionWithURL:card.url useWebView:card.useWebView];
  }
  [sender setTitle:@"Log Click ✅" forState:UIControlStateNormal];
}

- (void)logDismissal:(UIButton *)sender {
  [self.cards[sender.tag].context logDismissed];
  [sender setTitle:@"Log Dismissal ✅" forState:UIControlStateNormal];
  // logDismissed marks the card locally but does not immediately trigger the subscription
  // callback. In a production UI, you would also remove the card from your view here
  // rather than waiting for the next update.
}

+ (Section *)cardSectionFromCard:(BRZContentCardRaw *)card
                         atIndex:(NSInteger)index {
  NSMutableArray *fields = [NSMutableArray array];

  switch (card.type) {
  case BRZContentCardRawTypeClassic:
    [fields addObject:[Field fieldWithName:@"type" value:@"classic"]];
    break;
  case BRZContentCardRawTypeImageOnly:
    [fields addObject:[Field fieldWithName:@"type" value:@"banner"]];
    break;
  case BRZContentCardRawTypeCaptionedImage:
    [fields addObject:[Field fieldWithName:@"type" value:@"captionedImage"]];
    break;
  case BRZContentCardRawTypeControl:
    [fields addObject:[Field fieldWithName:@"type" value:@"control"]];
    break;
  default:
    break;
  }

  [fields addObject:[Field fieldWithName:@"id" value:card.identifier]];
  [fields addObject:[Field fieldWithName:@"image" value:card.image]];
  [fields addObject:[Field fieldWithName:@"imageAspectRation"
                                   value:@(card.imageAspectRatio)]];
  [fields addObject:[Field fieldWithName:@"title" value:card.title]];
  [fields addObject:[Field fieldWithName:@"description"
                                   value:card.cardDescription]];
  [fields addObject:[Field fieldWithName:@"domain" value:card.domain]];
  [fields addObject:[Field fieldWithName:@"url" value:card.url]];
  [fields addObject:[Field fieldWithName:@"useWebView"
                                   value:card.useWebView ? @"true" : @"false"]];
  [fields addObject:[Field fieldWithName:@"viewed"
                                   value:card.viewed ? @"true" : @"false"]];
  [fields addObject:[Field fieldWithName:@"dismissible"
                                   value:card.dismissible ? @"true" : @"false"]];
  [fields addObject:[Field fieldWithName:@"removed"
                                   value:card.removed ? @"true" : @"false"]];
  [fields addObject:[Field fieldWithName:@"pinned"
                                   value:card.pinned ? @"true" : @"false"]];
  [fields addObject:[Field fieldWithName:@"clicked"
                                   value:card.clicked ? @"true" : @"false"]];
  [fields addObject:[Field fieldWithName:@"test"
                                   value:card.test ? @"true" : @"false"]];
  [fields addObject:[Field fieldWithName:@"createdAt" value:@(card.createdAt)]];
  [fields addObject:[Field fieldWithName:@"expiresAt" value:@(card.expiresAt)]];

  NSData *extrasData = [NSJSONSerialization dataWithJSONObject:card.extras
                                                       options:0
                                                         error:nil];
  NSString *extrasString = [[NSString alloc] initWithData:extrasData
                                                 encoding:NSUTF8StringEncoding];
  [fields addObject:[Field fieldWithName:@"extras" value:@""]];
  [fields addObject:[Field fieldWithName:extrasString value:@"" indentation:1]];

  Section *section = [[Section alloc] init];
  section.name = [NSString stringWithFormat:@"Card %ld", index];
  section.fields = [fields copy];

  return section;
}

@end

#pragma mark - Field / Section Implementation

@implementation Field

- (instancetype)initWithName:(NSString *)name value:(id)value {
  self = [super init];
  if (self) {
    self.name = name;
    self.value = value;
  }
  return self;
}

+ (instancetype)fieldWithName:(NSString *)name value:(id)value {
  return [[Field alloc] initWithName:name value:value];
}

+ (instancetype)fieldWithName:(NSString *)name
                        value:(id)value
                  indentation:(NSInteger)indentation {
  Field *field = [Field fieldWithName:name value:value];
  field.indentation = indentation;
  return field;
}

@end

@implementation Section
@end

File: Examples/Swift/Sources/ContentCards-Custom-UI/CardsInfoViewController.swift
import BrazeKit
import UIKit

final class CardsInfoViewController: UITableViewController {

  // Represents a card property
  struct Field {
    let name: String
    var value: Any
    var indentation: Int = 0
  }

  // Represents a table view section
  struct Section {
    let name: String?
    var fields: [Field]
  }

  // MARK: - Properties

  let braze: Braze
  private var cards: [Braze.ContentCard] = []
  private var sections: [Section] = []

  // The subscription must be retained to keep it active.
  // Cancelling it re-applies local analytics (viewed/clicked/removed) and notifies other subscribers.
  private var subscription: Braze.Cancellable?

  // Tracks cards already impressed in this presentation to avoid duplicate impression events.
  private var impressedCardIDs: Set<String> = []

  // MARK: - Initialization

  init(braze: Braze) {
    self.braze = braze
    super.init(style: .grouped)
    title = "Content Cards Info"
    let doneButton = UIBarButtonItem(
      barButtonSystemItem: .done, target: self, action: #selector(dismissModal))
    navigationItem.setRightBarButton(doneButton, animated: false)

    tableView.rowHeight = UITableView.automaticDimension
    tableView.estimatedRowHeight = 44.0
  }

  @available(*, unavailable)
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  // MARK: - Lifecycle

  override func viewDidLoad() {
    super.viewDidLoad()
    // Subscribe to content cards updates. Delivers an initial snapshot of cached cards
    // immediately, then subsequent updates as cards change. Cancelling re-applies local
    // analytics (viewed/clicked/removed) and notifies any remaining subscribers.
    // Note: impressions, clicks, and dismissals do not trigger the subscription callback
    subscription = braze.contentCards.subscribeToUpdates { [weak self] cards in
      guard let self else { return }
      self.cards = cards
      self.sections = cards.enumerated().map { Self.cardSection(from: $1, index: $0) }
      self.tableView.reloadData()
    }
  }

  @objc
  func dismissModal() {
    dismiss(animated: true)
  }

  // MARK: - UITableViewDataSource

  override func numberOfSections(in tableView: UITableView) -> Int {
    sections.count
  }

  override func tableView(
    _ tableView: UITableView,
    numberOfRowsInSection section: Int
  ) -> Int {
    sections[section].fields.count
  }

  override func tableView(
    _ tableView: UITableView,
    cellForRowAt indexPath: IndexPath
  ) -> UITableViewCell {
    let cell =
      tableView.dequeueReusableCell(withIdentifier: "cellIdentifier")
      ?? UITableViewCell(style: .value1, reuseIdentifier: "cellIdentifier")

    let field = sections[indexPath.section].fields[indexPath.row]
    cell.textLabel?.text = field.name
    cell.detailTextLabel?.text = "\(field.value)"
    cell.indentationLevel = field.indentation

    cell.textLabel?.numberOfLines = 0
    if #available(iOS 13.0, tvOS 13.0, *) {
      cell.textLabel?.font = .monospacedSystemFont(ofSize: 13, weight: .regular)
      cell.detailTextLabel?.font = .monospacedSystemFont(ofSize: 13, weight: .regular)
    }

    return cell
  }

  override func tableView(
    _ tableView: UITableView,
    titleForHeaderInSection section: Int
  ) -> String? {
    sections[section].name
  }

  override func tableView(
    _ tableView: UITableView,
    viewForFooterInSection section: Int
  ) -> UIView? {
    let dismissButton = UIButton(type: .system)
    dismissButton.setTitle("Log Dismissal", for: .normal)
    dismissButton.tag = section
    dismissButton.addTarget(self, action: #selector(logDismissal(_:)), for: .touchUpInside)

    let clickButton = UIButton(type: .system)
    clickButton.setTitle("Log Click", for: .normal)
    clickButton.tag = section
    clickButton.addTarget(self, action: #selector(logClick(_:)), for: .touchUpInside)

    let stack = UIStackView(arrangedSubviews: [clickButton, dismissButton])
    stack.axis = .vertical
    stack.spacing = 8

    let container = UIView()
    container.addSubview(stack)
    stack.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      stack.centerXAnchor.constraint(equalTo: container.centerXAnchor),
      stack.topAnchor.constraint(equalTo: container.topAnchor, constant: 8),
      stack.bottomAnchor.constraint(equalTo: container.bottomAnchor, constant: -8),
    ])
    return container
  }

  // Log an impression when the first row of each card section becomes visible.
  // Deduplication prevents a second impression if the card scrolls off-screen and back.
  override func tableView(
    _ tableView: UITableView,
    willDisplay cell: UITableViewCell,
    forRowAt indexPath: IndexPath
  ) {
    guard indexPath.row == 0 else { return }
    let card = cards[indexPath.section]
    guard !impressedCardIDs.contains(card.id) else { return }
    impressedCardIDs.insert(card.id)
    card.context?.logImpression()
  }

  @objc
  func logClick(_ sender: UIButton) {
    let card = cards[sender.tag]
    card.context?.logClick()
    if let clickAction = card.clickAction {
      card.context?.processClickAction(clickAction)
    }
    sender.setTitle("Log Click ✅", for: .normal)
  }

  @objc
  func logDismissal(_ sender: UIButton) {
    cards[sender.tag].context?.logDismissed()
    sender.setTitle("Log Dismissal ✅", for: .normal)
    // logDismissed marks the card locally but does not immediately trigger the subscription
    // callback. In a production UI, you would also remove the card from your view here
    // rather than waiting for the next update.
  }

  // MARK: - Helpers

  static func cardSection(from card: Braze.ContentCard, index: Int) -> Section {
    var section = Section(name: "Card \(index)", fields: [])

    var type = Field(name: "type", value: "")
    var cardFields: [Field] = []
    switch card {
    case .classic(let classic):
      type.value = "classic"
      cardFields += [
        Field(name: "title", value: classic.title),
        Field(name: "description", value: classic.description),
        Field(name: "domain", value: classic.domain ?? "nil"),
      ]
    case .classicImage(let classicImage):
      type.value = "classicImage"
      cardFields += [
        Field(name: "image", value: classicImage.image),
        Field(name: "title", value: classicImage.title),
        Field(name: "description", value: classicImage.description),
        Field(name: "domain", value: classicImage.domain ?? "nil"),
      ]
    case .imageOnly(let imageOnly):
      type.value = "imageOnly"
      cardFields += [
        Field(name: "image", value: imageOnly.image),
        Field(name: "imageAspectRatio", value: imageOnly.imageAspectRatio ?? "nil"),
      ]
    case .captionedImage(let captionedImage):
      type.value = "captionedImage"
      cardFields += [
        Field(name: "image", value: captionedImage.image),
        Field(name: "imageAspectRatio", value: captionedImage.imageAspectRatio ?? "nil"),
        Field(name: "title", value: captionedImage.title),
        Field(name: "description", value: captionedImage.description),
        Field(name: "domain", value: captionedImage.domain ?? "nil"),
      ]
    case .control:
      type.value = "control"
    @unknown default:
      break
    }
    section.fields += [type] + cardFields

    section.fields += [Field(name: "id", value: card.id)]
    section.fields += fields(from: card.clickAction)
    section.fields += [
      Field(name: "viewed", value: card.viewed),
      Field(name: "dismissible", value: card.dismissible),
      Field(name: "removed", value: card.removed),
      Field(name: "pinned", value: card.pinned),
      Field(name: "clicked", value: card.clicked),
      Field(name: "test", value: card.test),
      Field(name: "createdAt", value: card.createdAt),
      Field(name: "expiresAt", value: card.expiresAt),
    ]

    return section
  }

  static func fields(
    from clickAction: Braze.ContentCard.ClickAction?,
    indentation: Int = 0
  ) -> [Field] {
    var fields: [Field] = []

    // Click Action
    var header = Field(name: "clickAction", value: "", indentation: indentation)
    switch clickAction {
    case .none:
      header.value = "none"
    case .url(let url, let useWebView):
      header.value = "url"
      fields = [
        Field(name: "url", value: url, indentation: indentation + 1),
        Field(name: "useWebView", value: useWebView, indentation: indentation + 1),
      ]
    @unknown default:
      break
    }

    return [header] + fields
  }

}
